### PR TITLE
Avoid using example/test courses in question sharing tests

### PR DESCRIPTION
**Pull request by nwalters512**
This PR switched the question sharing tests to use two dynamically-constructed courses instead of using the example course or test course. This is necessary for #9604, which will disallow consuming questions shared from the example course.

I generally like the pattern of creating new courses on the fly. That allows the configuration of the relevant pieces to be visible within the test file, and it means that we can't accidentally break tests by changing the example/test courses. Obviously, this wouldn't work as well when we actually want to test the contents of questions, or generally any non-JSON file in courses.

### CODE REVIEW

**Review by mwest1066 — APPROVED** [review]

**Comment by mwest1066** [comment]
> I generally like the pattern of creating new courses on the fly. That allows the configuration of the relevant pieces to be visible within the test file, and it means that we can't accidentally break tests by changing the example/test courses.

Agreed. It does however mean that you can't easily go and play with the test environment in a running server, but on balance I think it's a win.

File: apps/prairielearn/src/tests/questionSharing.test.ts [hunk 1]
```diff
@@ -103,62 +163,66 @@ describe('Question Sharing', function () {
...
+        CONSUMING_COURSE_SHARING_NAME,
+      );
     });
 
+    // TODO: fix this test?
     step('Fail if trying to set sharing name again.', async () => {
-      const result = await setSharingName(testCourseId, testCourseSharingName);
+      const result = await setSharingName(consumingCourse.id, CONSUMING_COURSE_SHARING_NAME);
       assert.equal(result.status, 200);
     });
```
Comment (SethPoulsen) [inline apps/prairielearn/src/tests/questionSharing.test.ts:212]: Can you be more specific in your comment? what about this test do you want to fix?

File: apps/prairielearn/src/tests/sync/util.ts [hunk 2]
```diff
@@ -470,12 +472,15 @@ export async function createAndSyncCourseData() {
...
-export async function writeAndSyncCourseData(courseData: CourseData): Promise<string> {
+export async function writeAndSyncCourseData(courseData: CourseData): Promise<{
+  courseDir: string;
+  syncResults: syncFromDisk.SyncResults;
+}> {
   const courseDir = await writeCourseToTempDirectory(courseData);
-  await syncCourseData(courseDir);
-  return courseDir;
+  const syncResults = await syncCourseData(courseDir);
+  return { courseDir, syncResults };
```
Comment (SethPoulsen) [inline apps/prairielearn/src/tests/sync/util.ts:483]: This is nice, I don't know why I didn't think to do this when I was writing these tests in the first place.

File: apps/prairielearn/src/tests/questionSharing.test.ts [hunk 3]
```diff
@@ -1,38 +1,39 @@
...
+import { selectCourseById } from '../models/course';
 import * as syncFromDisk from '../sync/syncFromDisk';
-
+import { fetchCheerio } from './helperClient';
+import * as helperServer from './helperServer';
 import { makeMockLogger } from './mockLogger';
+import * as syncUtil from './sync/util';
+
+const sql = sqldb.loadSqlEquiv(__filename);
 const { logger } = makeMockLogger();
```
Comment (SethPoulsen) [inline apps/prairielearn/src/tests/questionSharing.test.ts:19]: why all these changes? are we using some new import sorting method or something?

File: apps/prairielearn/src/tests/questionSharing.test.ts [hunk 4]
```diff
@@ -47,26 +48,85 @@ async function setSharingName(courseId, name) {
...
+  // from scratch. This is necessary because the example course has hardcoded behavior
+  // to prevent all question sharing features from working.
+  let sharingCourse: Course;
+  let consumingCourse: Course;
+  before('construct and sync course', async () => {
+    const sharingCourseData = syncUtil.getCourseData();
+    sharingCourseData.course.name = 'SHARING 101';
+    sharingCourseData.questions = {
+      [SHARING_QUESTION_QID]: {
+        uuid: '00000000-0000-0000-0000-000000000000',
```
Comment (SethPoulsen) [inline apps/prairielearn/src/tests/questionSharing.test.ts:81]: why not just generate a random UUID here and below? That seems more future-proof to me.

File: apps/prairielearn/src/tests/questionSharing.test.ts (hunk 1 above)
Comment (nwalters512) [reply apps/prairielearn/src/tests/questionSharing.test.ts:212]: The test is named "Fail...", but it asserts a 200 response. Either the title or the assertion (or the behavior?) is wrong.

File: apps/prairielearn/src/tests/sync/util.ts (hunk 2 above)
Comment (nwalters512) [reply apps/prairielearn/src/tests/sync/util.ts:483]: I think this is the first time we're using these utils outside of the syncing test, so I don't blame you for not thinking of it. I didn't either!

File: apps/prairielearn/src/tests/questionSharing.test.ts (hunk 4 above)
Comment (nwalters512) [reply apps/prairielearn/src/tests/questionSharing.test.ts:81]: 🤷 I don't think it matters much either way?

File: apps/prairielearn/src/tests/questionSharing.test.ts (hunk 3 above)
Comment (nwalters512) [reply apps/prairielearn/src/tests/questionSharing.test.ts:19]: We don't yet have a lint rule to enforce this (so I don't say anything on PRs), but yes, I try to keep first-party and third-party inputs separate, and I opportunistically clean that up in existing files when I can.

**Pull request merged**